cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp

```diff
@@ -391,3 +391,56 @@ void test_wcsdup(wchar_t *source)
 	sink(a); // tainted
 	sink(b);
 }
+
+// --- qualifiers ---
+
+class MyClass2 {
+public:
+	void setMember(int value);
+	int getMember();
+
+	int member;
+};
+
+class MyClass3 {
+public:
+	void setString(const char *string);
+	const char *getString();
+
+	const char *buffer;
+};
+
+void test_qualifiers()
+{
+	MyClass2 a, b, *c;
+	MyClass3 d;
+
+	sink(a); // [FALSE POSITIVE]
+	sink(a.getMember()); // [FALSE POSITIVE]
```

Comment:
I don't think these are false positives. They show that uninitialized data from `a` reach the sink, and they appear because flow from uninitialized data is part of the configuration for these tests: https://github.com/Semmle/ql/blob/bc88c41e0b611caf9501d476e77da61a3d8efba7/cpp/ql/test/library-tests/dataflow/dataflow-tests/DataflowTestCommon.qll#L29-L30

We should probably just remove that rule like I removed it for IR data flow in https://github.com/Semmle/ql/pull/2569/commits/64c79bf9e1e5ceaf4dadd24698f72aae11e51dc8, but that's independent of this PR.

Reply:
Yep, you're quite right and this was the cause of my problems.

In addition, I've figured out that flow to the qualifier needs to be into the partial definition of that qualifier.

I'm about to push a significantly reworked version.